29. Use multi-stage builds to build Docker images for microservices

@sergey-morenets I have an Issue here

dockerfile:

FROM maven:3-openjdk-18 as maven

COPY pom.xml /opt/

WORKDIR /opt

RUN mvn dependency:resove

COPY . /opt/

RUN mvn install spring-boot:repackage

FROM openjdk

LABEL Author="Michał Tworuszka, Oskar Hanke, Paweł Manowski"

#ARG JAR_FILE=target/*.jar
COPY --from=maven  /opt/target/game_shop-notifications-0.0.1-SNAPSHOT.jar notifications.jar

WORKDIR /opt

ENTRYPOINT ["java","-jar","notifications.jar"]




But when I build it I have error:

 > [notifications maven 4/6] RUN mvn dependency:resove:
#0 1.459 [INFO] Scanning for projects...
#0 1.486 [ERROR] [ERROR] Some problems were encountered while processing the POMs:
#0 1.486 [FATAL] Non-resolvable parent POM for org.springframework.boot:game_shop-notifications:0.0.1-SNAPSHOT: Could not find artifact org.springframework.boot:game_shop:pom:0.0.1-SNAPSHOT and 'parent.relativePath' points at wrong 
local POM @ line 5, column 13
#0 1.486  @
#0 1.488 [ERROR] The build could not read 1 project -> [Help 1]
#0 1.490 [ERROR]
#0 1.490 [ERROR]   The project org.springframework.boot:game_shop-notifications:0.0.1-SNAPSHOT (/opt/pom.xml) has 1 error
#0 1.491 [ERROR]     Non-resolvable parent POM for org.springframework.boot:game_shop-notifications:0.0.1-SNAPSHOT: Could not find artifact org.springframework.boot:game_shop:pom:0.0.1-SNAPSHOT and 'parent.relativePath' points at wr
ong local POM @ line 5, column 13 -> [Help 2]
#0 1.491 [ERROR]
#0 1.493 [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
#0 1.494 [ERROR] Re-run Maven using the -X switch to enable full debug logging.
#0 1.494 [ERROR]
#0 1.494 [ERROR] For more information about the errors and possible solutions, please read the following articles:
#0 1.494 [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/ProjectBuildingException
#0 1.494 [ERROR] [Help 2] http://cwiki.apache.org/confluence/display/MAVEN/UnresolvableModelException
------
failed to solve: executor failed running [/bin/sh -c mvn dependency:resove]: exit code: 1





On Lectures you had the same error and you "deleted" from child pom parrent "dependency"

But can I do that? Because I have some dependencies and dependencie manager in Parrent pom.

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,13 +1,59 @@
+FROM maven:3-openjdk-18 as builder
+
+ENV HOME=/usr/app
+RUN mkdir -p $HOME
+WORKDIR $HOME
+
+COPY . $HOME
+
+RUN mvn clean install
+
+WORKDIR $HOME/eureka-server
+RUN mvn clean install spring-boot:repackage
+
+WORKDIR $HOME/config-server
+RUN mvn clean install spring-boot:repackage
+
+WORKDIR $HOME/game_shop-orders-client
+RUN mvn clean install
+
+WORKDIR $HOME/game_shop-users-client
+RUN mvn clean install
+
+WORKDIR $HOME/game_shop-notifications
+RUN mvn clean install spring-boot:repackage
+
+WORKDIR $HOME/game_shop-orders
+RUN mvn clean install spring-boot:repackage
+
+WORKDIR $HOME/game_shop-users
+RUN mvn clean install spring-boot:repackage
+
+
 FROM openjdk:18-jdk-alpine
+ENV HOME=/usr/app
+RUN mkdir -p $HOME
+WORKDIR $HOME
 
-LABEL Author="Michał Tworuszka, Oskar Hanke, Paweł Manowski"
+#LABEL Author="Michał Tworuszka, Oskar Hanke, Paweł Manowski"
 
-EXPOSE 8080
-ARG JAR_FILE=target/*.jar
-COPY ${JAR_FILE} app.jar
-ENTRYPOINT ["java","-jar","/app.jar"]
+COPY --from=builder $HOME/eureka-server/target/eureka-server-0.0.1-SNAPSHOT.jar eureka-server.jar
 
+ENTRYPOINT ["java","-jar","/eureka-server.jar"]
 
+COPY --from=builder $HOME/config-server/target/config-server-0.0.1-SNAPSHOT.jar config-server.jar
 
+ENTRYPOINT ["java","-jar","/config-server.jar"]
 
+COPY --from=builder $HOME/game_shop-notifications/target/game_shop-notifications-0.0.1-SNAPSHOT.jar notifications.jar
+
+ENTRYPOINT ["java","-jar","notifications.jar"]
+
+COPY --from=builder $HOME/game_shop-orders/target/game_shop-orders-0.0.1-SNAPSHOT.jar orders.jar
+
+ENTRYPOINT ["java","-jar","orders.jar"]
+
+COPY --from=builder $HOME/game_shop-users/target/game_shop-users-0.0.1-SNAPSHOT.jar users.jar
+
+ENTRYPOINT ["java","-jar","users.jar"]
 

--- a/game_shop-orders/pom.xml
+++ b/game_shop-orders/pom.xml
@@ -12,6 +12,10 @@
 
     <artifactId>game_shop-orders</artifactId>
 
+    <properties>
+        <maven.test.skip>true</maven.test.skip>
+    </properties>
+
     <dependencies>
         <dependency>
             <groupId>org.springframework.boot</groupId>

--- a/game_shop-users/pom.xml
+++ b/game_shop-users/pom.xml
@@ -12,8 +12,7 @@
     <artifactId>game_shop-users</artifactId>
 
     <properties>
-        <maven.compiler.source>18</maven.compiler.source>
-        <maven.compiler.target>18</maven.compiler.target>
+        <maven.test.skip>true</maven.test.skip>
     </properties>
 
     <dependencies>

--- a/pom.xml
+++ b/pom.xml
@@ -53,6 +53,7 @@
 		<dependency>
 			<groupId>org.projectlombok</groupId>
 			<artifactId>lombok</artifactId>
+			<scope>provided</scope>
 		</dependency>
 		<dependency>
 			<groupId>org.springframework.boot</groupId>

